File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

bool visited[201];

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    for(int i = 0; i < n; ++i){
        if(visited[i]) continue;
        
        answer++;
        queue<int> q;
        q.push(i);
        visited[i] = true;
        
        while(!q.empty()){
            int tmp = q.front();
            q.pop();
            
            for(int j = 0; j < n; ++j){
                if(visited[j] || !computers[tmp][j]) continue;
                visited[j] = true;
                q.push(j);
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
class Solution {
    boolean[] visited;
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        visited = new boolean[n];
        
        for(int i = 0; i < n; i++) {
            if(!visited[i]) {
                dfs(n, computers, i);   
                answer++;
            }
        }
        
        return answer;
    }
    
    void dfs(int n, int[][] computers, int x) {
        visited[x] = true;
        
        for(int i = 0; i < n; i++) {
            if(i == x || visited[i]) continue;
            
            if(computers[i][x] == 1) {
                dfs(n, computers, i);
            }
        }
    }
}